ROX-14549: Explaining Red Hat SSO configuration in ACS Console (#56937)
openshift/openshift-docs: operating/manage-user-access/authentication-provider-structure.adoc

[id="auth-provider-claim-mappings"]
== Claim mappings
Claim mappings allow customer to specify if attribute in user identity obtained from IdP should be added to ACS-issued token and with which attribute name.
Example of claim mapping: from `roles` in user identity to `groups` in ACS-issued token.
There are different default claim mappings for each auth provider type.

[id="auth-provider-oidc-default-claim-mappings"]
=== OIDC default claim mappings
* `sub` to `userid`
* `name` to `name`
* `email` to `email`
* `groups` to `groups`

[id="auth-provider-auth0-default-claim-mappings"]
=== Auth0 default claim mappings
Same as OIDC default claim mappings.

[id="auth-provider-saml20-default-claim-mappings"]
=== SAML 2.0 default claim mappings
* `Subject.NameID` to `userid`
* every SAML `AttributeStatement.Attribute` from response gets mapped to its name

[id="auth-provider-google-iap-default-claim-mappings"]
=== Google IAP default claim mappings
* `sub` to `userid`
* `email` to `email`
* `hd` to `hd`
* `google.access_levels` to `access_levels`

[id="auth-provider-userpki-default-claim-mappings"]
=== User Certificates default claim mappings

> User Certificates is different from all other authentication providers in a sense that instead of communicating with a third-party IdP, it get user information from certificate used by user.

Reviewer: ```suggestion
User certificates differ from all other authentication providers because instead of communicating with a third-party IdP, they get user information from certificates used by the user.
```

> Default claim mappings:

Reviewer: ```suggestion
The default claim mappings for user certificates include:
```

* `CertFingerprint` to `userid`
* `Subject -> Common Name` to `name`
* `EmailAddresses` to `email`
* `Subject -> Organizational Unit` to `groups`

> 

Reviewer: remove blank line

[id="auth-provider-openshift-auth-default-claim-mappings"]
=== Openshift Auth default claim mappings
* `groups` to `groups`
* `uid` to `userid`
* `name` to `name`

[id="auth-provider-rules"]
== Rules

> To authorize users, RHACS relies on mapping rules that examine the values of specific claims like, `groups`, `email`, `userid` and `name` from user identity.

Reviewer: ```suggestion
To authorize users, {product-title-short} relies on mapping rules that examine the values of specific claims such as `groups`, `email`, `userid`, and `name` from the user identity.
```

Rules allow to map users who have attribute with a specific value to a specific role. Example of a rule: `key` is `email`, `value` is `john@redhat.com`, `role` is `Admin`.

If claim is absent, the mapping cannot succeed and the user does not get access to the required resources.
Therefore, you need to ensure that the claims required to authorize users from your IdP, for example, `groups`, are included in the authentication response of your IdP to enable successful mapping.

[id="auth-provider-minimum-access-role"]
== Minimum access role
Minimum access role is role that all is assigned to every caller with ACS token issued by particular authentication provider. It is set to `None` by default.

For example, if there is authentication provider with minimum access role `Analyst`, all users who logged in using this provider will have role `Analyst` assigned to them.

[id="auth-provider-required-attributes"]
== Required attributes
Required attributes allow to restrict issuing of the ACS token based on whether user identity has attribute with a specific value.

> Example of required attribute: `key` is `is_internal`, `value` is `true`. In that case users with `is_internal` not set or set to `false` will not be logged in.

Reviewer: ```suggestion
For example, when `key` is `is_internal` and `value` is `true`, users with `is_internal` not set or set to `false` will not be logged in.
```

:_content-type: ASSEMBLY
[id="authentication-provider-structure"]
= Authentication provider structure
include::modules/common-attributes.adoc[]
:context: authentication-provider-structure

toc::[]

[role="_abstract"]

> An authentication provider connects to a third-party source of user identity (for example, an identity provider, or IDP). An authentication provider gets the user identity from IDP, issues a token based on this identity, and returns the token to the caller. The mentioned token is then sent within the user interface or API calls and allows {product-title-short} to authorize the user.

Reviewer 2: ```suggestion
An authentication provider connects to a third-party source of user identity, for example, an identity provider or IDP, gets the user identity, issues a token based on that identity, and returns the token to RHACS. 
This token allows RHACS to authorize the user. 
RHACS uses the token within the user interface and API calls.

After installing RHACS, you must set up your IDP to authorize users. 
```

> Navigate to `Operating -> Managing user access`, and select the appropriate module to set up your own IDP immediately after provisioning.

Reviewer 2: **issue:** I don't think this is pointing to RHACS UI, instead a section in the docs. Additionally, this PR changes the section placement. 

To fix this, we'll need to add `xref` pointing to existing content based on new TOC.

> You can set up your own IdP after provisioning by navigating to *Operating* -> *Managing user access* and selecting the appropriate module. 

Reviewer 3: This sentence is in passive voice. 

s/ Navigate to `Operating -> Managing user access`, and select the appropriate module to set up your own IDP immediately after provisioning.

[id="auth-provider-claim-mappings"]
== Claim mappings

> Claim mappings allow a customer to specify if an attribute in the user identity obtained from the IDP should be added to the {product-title-short}-issued token and with which attribute name.

Reviewer 2: ```suggestion
A claim is the data an identity provider includes about a user inside the token they issue. 

Using claim mappings, you can specify if RHACS should customize the claim attribute it receives from an IDP to another attribute in the {product-title-short}-issued token. 

For example, you can map from `roles` in the user identity to `groups` in the {product-title-short}-issued token using claim mapping.

RHACS uses different default claim mappings for every authentication provider.
```

> Claim mappings allow a customer to specify if an attribute in the user identity obtained from the IDP should be added to the {product-title-short}-issued token and with which attribute name.
> An example of claim mapping is the mapping from `roles` in the user identity to `groups` in the {product-title-short}-issued token.
> There are different default claim mappings for each auth provider type.

Reviewer 2: ```suggestion
A claim is the data an identity provider includes about a user inside the token they issue. 

Using claim mappings, you can specify if RHACS should customize the claim attribute it receives from an IDP to another attribute in the {product-title-short}-issued token. 

For example, you can map from `roles` in the user identity to `groups` in the {product-title-short}-issued token using claim mapping.

RHACS uses different default claim mappings for every authentication provider.
```

Author: The point that would be important to make is that without claim mapping, a claim from IdP will not be added to the ACS-issued token.

Reviewer 2: Thank you @ivan-degtiarenko we can add the following after the `Using claim mapping ...` sentence:
```
If you do not use the claim mapping, {product-title-short} does not include the claim attribute in the {product-title-short}-issued token.
```

* `email` to `email`
* `hd` to `hd`
* `google.access_levels` to `access_levels`

[id="auth-provider-userpki-default-claim-mappings"]
=== User Certificates default claim mappings
User certificates differ from all other authentication providers because instead of communicating with a third-party IDP, they get user information from certificates used by the user.

The default claim mappings for user certificates include:

* `CertFingerprint` to `userid`
* `Subject -> Common Name` to `name`
* `EmailAddresses` to `email`
* `Subject -> Organizational Unit` to `groups`

[id="auth-provider-openshift-auth-default-claim-mappings"]
=== OpenShift Auth default claim mappings
The following list provides the OpenShift Auth default claim mappings:

* `groups` to `groups`
* `uid` to `userid`
* `name` to `name`

[id="auth-provider-rules"]
== Rules
To authorize users, {product-title-short} relies on mapping rules that examine the values of specific claims such as `groups`, `email`, `userid`, and `name` from the user identity.
Rules allow to map users who have attribute with a specific value to a specific role. Example of a rule: `key` is `email`, `value` is `john@redhat.com`, `role` is `Admin`.

> If the claim is absent, the mapping cannot succeed and the user does not get access to the required resources.
> Therefore, you need to ensure that the claims required to authorize users from your IDP (for example, `groups`) are included in the authentication response of your IDP to enable successful mapping.

Reviewer 2: ```suggestion
If the claim is missing, the mapping cannot succeed, and the user does not get access to the required resources.
Therefore, to enable successful mapping, you must ensure that the authentication response from your IDP includes the required claims to authorize users, for example, `groups`.
```

[id="auth-provider-minimum-access-role"]
== Minimum access role

> The minimum access role is the role that all is assigned to every caller with an {product-title-short} token issued by a particular authentication provider. It is set to `None` by default.
> 
> For example, if there is an authentication provider with the minimum access role of `Analyst`, all users who log in using this provider have the role of `Analyst` assigned to them.

Reviewer 2: ```suggestion
{product-title-short} assigns a minimum access role to every caller with a {product-title-short} token issued by a particular authentication provider. The minimum access role is set to `None` by default.

For example, suppose there is an authentication provider with the minimum access role of `Analyst`. In that case, all users who log in using this provider will have the `Analyst` role assigned to them.
```

[id="auth-provider-required-attributes"]
== Required attributes
Required attributes can restrict issuing of the {product-title-short} token based on whether a user identity has an attribute with a specific value.

> For example, when `key` is `is_internal` and `value` is `true`, users with `is_internal` not set or set to `false` will not be logged in.

Reviewer 2: May be:
```
Required attributes can restrict issuing of the {product-title-short} token based on whether a user identity has an attribute with a specific value.

For example, you can configure RHACS only to issue a token when the attribute `key` has the value `is_internal` and the attribute `value` has the value `true`. 
Users with the attribute `key` with the value `is_internal` and the attribute `value` with the value `false` do not get a token.
```

Author: Attributes(and claims) have key and then value(for example, attribute can have key `email` and value `john@redhat.com`). Tried to rewrite based on your suggestion

:_content-type: ASSEMBLY
[id="authentication-provider-structure"]

> = Authentication provider structure

Reviewer 3: ```suggestion
= Understanding authentication providers
```

include::modules/common-attributes.adoc[]
:context: authentication-provider-structure

toc::[]

[role="_abstract"]

> An authentication provider connects to a third-party source of user identity, for example, an identity provider or IDP, gets the user identity, issues a token based on that identity, and returns the token to RHACS. 

Reviewer 3: ```suggestion
An authentication provider connects to a third-party source of user identity, for example, an identity provider (IDP), gets the user identity, issues a token based on that identity, and returns the token to {rh-rhacs-first}. 
```

Reviewer 3: Dictionary: https://source.redhat.com/groups/public/red-hat-dictionary-or-lexicon/red_hat_abbreviations_acronyms_and_initialisms_dictionary_wiki/rh_dictionary_letter_i#:~:text=idP,Identity%20Provider

Author: this contradicts what @kcarmichael08 proposed :(

> This token allows RHACS to authorize the user. 

Reviewer 3: ```suggestion
This token allows {product-title-short} to authorize the user. 
```

> RHACS uses the token within the user interface and API calls.

Reviewer 3: ```suggestion
{project-title-short} uses the token within the user interface and API calls.
```

> RHACS uses the token within the user interface and API calls.

Reviewer 3: ```suggestion
{project-title-short} uses the token within the user interface and API calls.
```

> {project-title-short} uses the token within the user interface and API calls.

Reviewer 3: ```suggestion
{product-title-short} uses the token within the user interface and API calls.
```

> After installing RHACS, you must set up your IDP to authorize users. 

Reviewer 3: ```suggestion
After installing RHACS, you must set up your idP to authorize users. 
```

Reviewer: @agantony @ivan-degtiarenko - IDP is correct per Red Hat Supplementary Style Guide.

Navigate to *Operating -> Managing user access*, and select the appropriate module to set up your own IDP immediately after provisioning.

> 

Reviewer 3: ```suggestion

[IMPORTANT]
====
If you are using OpenID Connect (OIDC) as an idP, {product-title-short} relies on mapping rules that examine the values of specific claims like `groups`, `email`, `userid` and `name` from either the user ID token or the `UserInfo` endpoint response to authorize the users.
====

If these details are absent, the mapping cannot succeed and the user does not get access to the required resources.
Therefore, you need to ensure that the claims required to authorize users from your IDP, for example, `groups`, are included in the authentication response of your IDP to enable successful mapping.

.Example

To connect an Azure Active Directory (AD) to {product-title-short} using  Sign-On (SSO) configuration, you need to add specific claims, for example, `groups` claim to tokens, and assign users, groups, or both to the enterprise application.
For instructions, see link:https://learn.microsoft.com/en-us/azure/active-directory/hybrid/how-to-connect-fed-group-claims#add-group-claims-to-tokens-for-saml-applications-using-sso-configuration[Add group claims to tokens for SAML applications using SSO configuration].
```

Author: To be honest I think the Azure example, even though it is a short one, should go on a separate page under `Configuring auth providers`. This would allow for better discoverability compared to

Reviewer 3: ok, SGTM!

Please add the below content to your PR: 

> :_content-type: ASSEMBLY
> [id="connecting-azure-ad-rhacs-using-sso-configuration"]
> = Connecting Azure AD to {product-title-short} using SSO configuration
> include::modules/common-attributes.adoc[]
> :context: user-authentication
> 
> 
> [role="_abstract"]
> 
> To connect an Azure Active Directory (AD) to {product-title-short} using  Sign-On (SSO) configuration. you need to add specific claims, for example, `groups` claim to tokens, and assign users, groups, or both to the enterprise application.
> 
> 
> [id="adding-group-claims-to-tokens-for-saml-applications-using-SSO-configuration_{context}"]
> == Adding group claims to tokens for SAML applications using SSO configuration
> 
> Configure the application registration in Azure AD to include group claims in tokens. For instructions, see link:https://learn.microsoft.com/en-us/azure/active-directory/hybrid/how-to-connect-fed-group-claims#add-group-claims-to-tokens-for-saml-applications-using-sso-configuration[Add group claims to tokens for SAML applications using SSO configuration].

NOTE: I added a suggestion to update the topic.yaml file.

An authentication provider connects to a third-party source of user identity (for example, an identity provider, or IdP). An authentication provider gets the user identity from IdP, issues a token based on this identity, and returns the token to the caller. The mentioned token is then sent within the user interface or API calls and allows {product-title-short} to authorize the user.

> Navigate to Operating -> Managing user access, and select the appropriate module to set up your own IDP immediately after provisioning.

Reviewer 3: `Operating -> Managing user access`

Reviewer: Per Style Guide, I think this is bold and not backticks. Backticks are normally for objects or items and this is a menu choice on the web page (unless I'm misunderstanding).
